pkg/yamltemplate/filetests/ytt-library/url.yml

```
@@ -19,6 +19,21 @@ query_params:
   test3: #@ url.query_params_encode({"w":[]})
   test4: #@ url.query_params_decode("")
 
+url_type:
+  #@ testURL1 = url.parse("http://admin:password@url.domain:8080")
+  #@ testURL2 = testURL1.without_user()
+  #@ testURL3 = url.parse("http://admin@url.domain:8080")
+  test1: #@ testURL1
+  test2: #@ testURL1.user
+  test3: #@ testURL1.user.name
+  test4: #@ testURL1.user.password
+  test5: #@ testURL2
+  test6: #@ testURL2.user
+  test7: #@ testURL3
+  test8: #@ testURL3.user
+  test9: #@ testURL3.user.name
+  test10: #@ testURL3.user.password
+
```

```suggestion
url_type:
  #@ testURL1 = url.parse("http://admin:password@url.domain:8080")
  #@ testURL2 = testURL1.without_user()
  #@ testURL3 = url.parse("http://admin@url.domain:8080")
  test1: #@ testURL1.string()
  test2: #@ testURL1.user
  test3: #@ testURL1.user.name
  test4: #@ testURL1.user.password
  test5: #@ testURL2.string()
  test6: #@ testURL2.user
  test7: #@ testURL3.string()
  test8: #@ testURL3.user
  test9: #@ testURL3.user.name
  test10: #@ testURL3.user.password

```

As noted in the summary comment, let's follow your initial idea of _not_ inferring a string encoding from a "naked" instance.

This means that there's a cryptic error message complaining "`...unknown type *starlark.Builtin for conversion to go value...`" 
... it's _this_ situation that makes me think we'd get to where we want to go faster and with the needed control if we "rolled our own" out the gate.

---

I think we want to make following changes after having a discussion last Friday:
1. Instead of using the legacy `starlarkstruct.Struct`, we want to converge to `StarlarkStruct`.
2. Instead of returning a confusing error message `...unknown type *starlark.Builtin for conversion to go value...` when referencing `testURL` directly, we want to see thing like `testURL cannot be directly referenced, please use testURL.string()`, something more instructive like that.
3. As @jtigger mentioned, involve a new method `string()` to return the string representation of the URL. We do this   1). to accommodate the potential use case that we might need to return a encoded string that contains unicode characters, 2). we might introduce more module-level types later on, and we might not always want to leave some flexibility for those types for the representations.

I'll try and see if it's feasible to address all these requirements in a single changeset.